Add Azure deployment cookbook (#1459)
meilisearch/documentation: learn/cookbooks/azure.md

# Deploy a Meilisearch instance on Azure

_The following is a guest post by Christopher Maneu, Data Engineering Lead Advocate at Microsoft._

## Introduction

In this tutorial, we will see how to deploy a Meilisearch instance on Azure, and also some considerations for hosting a production Meilisearch instance on Azure.

::: tip
If you don't have yet an Azure subscription, you can [create one here](https://azure.microsoft.com/free/). With the Azure Free tier, you can run a Meilisearch instance for free for at least 12 months.
:::

## One-Click Deploy

You can deploy a Meilisearch instance via the official [Meilisearch Docker image](https://hub.docker.com/r/getmeili/meilisearch) by clicking the following button:

[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fcmaneu%2Fmeilisearch-on-azure%2Fmain%2Fmain.json).

![The Azure portal prompting for information to deploy Meilisearch](/azure/01.azure-deploy-button.png)

> When clicking this button, you'll be redirected to the Azure Portal, and asked a few questions:

Reviewer: ```suggestion
When clicking this button, you'll be redirected to the Azure Portal and asked a few questions:
```

Author: Our internal style tool always yell at me when I don't put a comma before an and. So in English, I add them "everywhere". Fact is it's not [right](https://www.grammarly.com/blog/comma-before-and/). Thanks for this discovery ;)

> - **Region**: This is the deployment region. You should select a region close to you or your users,

Reviewer: ```suggestion
- **Region**: This is the deployment region. You should select a region close to you or your users.
```

- **Environment**: The name of the environment you want to deploy to, like `dev` (for development) or `prod` (for production). This settings has no effects on the type of resources deployed. It's just for convenience.
- **Application Name**: The name of your application. This name should be **unique across all Azure customers**. The URL where your application is deployed will contain this value.
- **Location**: This the resources location. For reasons out of scope of this documentation, this parameter is different from the _Region_ one. By default, it is set to `eastus`, but you can type any Azure location (like `francecentral`, `westeurope`, or `japaneast`).

> - **Meilisearch_apikey**: This is the Meilisearch Master Key. While we generate a random key by default, we encourage you to set your own key.

Reviewer: ```suggestion
- **Meilisearch_apikey**: This will be the [master key](/learn/advanced/security.md) of your Meilisearch instance. While we generate a random key by default, we encourage you to set your own key.
```

This code will setup two main components:

- An **Azure App Service** instance to host Meilisearch container
- An **Azure File Share** who is volume-mapped into the Meilisearch container for database persistence.

## Production-ready deployment

We've made some _opinionated choices_ in the One-click deployment that should work for most production-ready deployments. This section explains some of these choices you need to be aware in the context of a production deployment.

### Data redundancy and backup

The one-click button create a storage with the "LRS" redundancy option. That means that your data is replicated across multiples servers, but in the same Datacenter. You can change - by modifying the template) this setting to choose a multi-zone or a multi-region redundancy option.

By default, there is no automatic backup of your index. If you want to use Azure Backup capabilities to achieve this result, you can look at [Azure File Share Backup](https://docs.microsoft.com/azure/backup/azure-file-share-backup-overview).

::: tip

> Meilisearch has a snapshot and dump features that could help you achieve backup capabilities. See [Data Backup documentation](/learn/advanced/snapshots_vs_dumps.md) for more information.

Reviewer: ```suggestion
Meilisearch has snapshot and dump features that can help improve your backup capabilities. See [Data Backup documentation](/learn/advanced/snapshots_vs_dumps.md) for more information.
```

:::

### Always-on

> To enable free tier, the _Always-on_ option is disabled by default. That means that after some time without any traffic, your instance will be deallocated. When the next client hit your instance, it'll be started again automatically. However, you will have a response time in the dozen-seconds range.

Reviewer: ```suggestion
To enable free tier, the _Always-on_ option is disabled by default. That means that after some time without any traffic, your instance will be deallocated. When the next client hits your instance, it'll be started again automatically. However, you will have a response time in the dozen-seconds range.
```

You can avoid this _cold start_ by enabling the _Always-on_ feature, either in the Azure portal or in the Infrastructure as Code.

### App Service Auto-Scale

>  Azure App Service offer an horizontal auto-scale feature that allows to automatically add more servers to process incoming requests.

Reviewer: ```suggestion
 Azure App Service offers a horizontal auto-scale feature that allows you to automatically add more servers to process incoming requests.
```
